client/js/chart/index.js

```diff
@@ -0,0 +1,130 @@
+  seriesContainer.setAttribute('class', 'series-container');
+  svg.append(seriesContainer);
+
+  let index = 0;
+  const colors = ['blue', 'red', 'green', 'purple', 'orange'];
+
+  for (const [, history] of [...rankingHistory].slice(0, 7)) {
+    const color = colors[index % colors.length];
+    const series = createSeries(color, history);
+    seriesContainer.append(series);
+    index++;
+  }
+
+  el.append(svg);
+}
+
+const RUN_SCALE_X = 30;
+const RUN_SCALE_Y = 30;
+
+/**
+ * Creates a new series (set of lines on the chart for a specific song).
+ *
+ * @param {string} color The color of this series.
+ * @param {number[]} history The historical positions of this track on the
+ * leaderboard.
+ * @returns {SVGGElement} A group containing the series.
+ */
+function createSeries(color, history) {
+  const series = document.createElementNS(SVG_NS, 'g');
+  series.setAttribute('class', 'series');
+  series.style.setProperty('--run-color', color);
+  let start = 0;
+  let end = 0;
+  while (end < history.length) {
+    while (start < history.length && history[start] === null) {
+      start++;
+    }
+    end = start + 1;
+    while (end < history.length && history[end] !== null) {
+      end++;
+    }
+    series.append(...createRun(history, start, end));
+    start = end;
+  }
+
+  return series;
+}
+
+/**
+ * Creates a new run (continuous group of points within a series).
+ *
+ * @param {number[]} history The historical positions of this track on the
+ * leaderboard.
+ * @param {number} start The index of the first entry in this run.
+ * @param {number} end The index of the last entry in this run.
+ * @returns {SVGElement[]} The elements that compose this run.
+ */
+function createRun(history, start, end) {
+  const points = history
+      .slice(start, end)
+      .map((val, idx) => `${(idx + start) * RUN_SCALE_X},${val * RUN_SCALE_Y}`)
+      .join(' ');
+
+  // line that is displayed
+  const line = document.createElementNS(SVG_NS, 'polyline');
+  line.setAttribute('class', 'series-run');
+  line.setAttribute('points', points);
+
+  // secondary invisible line to make it easier to hit the line
+  // with the mouse
+  const touchTarget = document.createElementNS(SVG_NS, 'polyline');
+  touchTarget.setAttribute('class', 'series-run-touch-target');
+  touchTarget.setAttribute('points', points);
+
+  const startCap = document.createElementNS(SVG_NS, 'circle');
+  const endCap = document.createElementNS(SVG_NS, 'circle');
+
+  startCap.setAttribute('class', 'series-run-cap');
+  endCap.setAttribute('class', 'series-run-cap');
+
+  startCap.setAttribute('r', '5');
+  startCap.setAttribute('cx', start * RUN_SCALE_X + 'px');
+  startCap.setAttribute('cy', history[start] * RUN_SCALE_Y + 'px');
+  endCap.setAttribute('r', '5');
+  endCap.setAttribute('cx', (end - 1) * RUN_SCALE_X + 'px');
+  endCap.setAttribute('cy', history[end - 1] * RUN_SCALE_Y + 'px');
+
+  return [startCap, line, touchTarget, endCap];
```

Comment:
Can you elaborate please? I don't really get what is difficult to understand about this.

Reply:
I think seun is saying it makes more sense to call an object attribute/method to get `startCap`, `line`, `touchTarget`, `endCap` respectively, instead of getting indices 0-3 of this list. With a list, if you want `startCap` you'd call `list[0]`. This seems kind of like a magic number instead of calling a method/attribute `.getStartCap()` of the object

But since you're iterating through this list, on line 58, returning an object doesn't make sense.

Reply:
Can you instead wrap up each run in its own element and just return that? This will also make your svg easier to read/debug:
```
const container = document.createElementNS(SVG_NS, 'g');
container.append(startCap, line, touchTarget, endCap);
return container;
```